Added CosFace Loss

crsdvaibhav opened the pull request:
#### @monatis This is still work in progress. Just wanted to ask few questions, by sharing the code.
 - Is the implementation of the Multi Objective Loss according to your demands? If it is not, can you give a few pointers in the correct direction, and also about the tests?
 - CosFace Loss is correct according to me, as its only difference from Arcface Loss is the subtracting of margin from the cosines instead of angles. Could you give a pointer regarding writing the tests for the Cosface and Arcface losses?

Thankyou for the help!

KarahanS commented:
Hi @crsdvaibhav, there is already a `l2_norm` function implemented for `arcface_loss`. If you are going to use `l2_norm` for `cos_face_loss` as well, then maybe it would be better to take `l2_norm` function to `loss/utils/utils.py` and call it from there in order to avoid code duplication.

KarahanS commented on quaterion/loss/cos_face_loss.py:
```diff
@@ -0,0 +1,85 @@
+from quaterion.loss.group_loss import GroupLoss
+
+
+def l2_norm(inputs: torch.Tensor, dim: int = 0) -> torch.Tensor:
+    """Apply L2 normalization to tensor
+    Args:
+        inputs: Input tensor.
+        dim: Dimension to operate on.
+    Returns:
+        torch.Tensor: L2-normalized tensor
+    """
+    outputs = inputs / torch.norm(inputs, 2, dim, True)
+
+    return outputs
```
Hi @crsdvaibhav, there is already a `l2_norm` function implemented for `arcface_loss`. If you are going to use `l2_norm` for `cos_face_loss` as well, then maybe it would be better to take `l2_norm` function to `loss/utils/utils.py` and call it from there in order to avoid code duplication.

KarahanS commented on quaterion/loss/multi_objective_loss.py:
```diff
@@ -0,0 +1,22 @@
+
+
+class MultiObjectiveLoss(torch.nn.Module):
+    def __init__(self, losses: list, weights: Optional[list] = None):
```
I think, in order to improve expressivity, that would be better to include data type stored in the list such as `weights: Optional[List[Tensor]]`.

crsdvaibhav commented:
@KarahanS I have made the changes. Does this need some more changes before merge?

KarahanS commented:
> @KarahanS I have made the changes. Does this need some more changes before merge?

Thanks for the updates! I'm not one of the official contributors of this project, so let's wait for the reviews by the maintainers for merging :)

crsdvaibhav commented:
@monatis I have added the CosFace Loss and its test. Could you please review it? I have also added an implementation of Multi Objective Loss, could you please review it?

monatis commented:
I'll review this PR, but could you please fix the conflict meanwhile? @crsdvaibhav

monatis commented on quaterion/loss/multi_objective_loss.py:
```diff
@@ -0,0 +1,22 @@
+from typing import Optional
```
This is a good start, but it has some issues:

- Not well documented with docstrings.
- Argument `groups` is missing in the `forward()` method.
- Tests are missing.

After fixing these issue it might be a good start, but I'm thinking of something slightly different that will make `GroupLoss` and `PairLoss` compatible with one another through some utility functions. For example, I got slightly better results when I combined Triplet Loss and Multiple Negatives Ranking Loss, but it will require more experiments.

tl; dr: Let's remove `MultiObjectiveLoss` from this PR for now --it will be something more ambitious and definitely is worth its own PR 😁

crsdvaibhav replied:
Sure! I will update the PR with the CosFace Loss only. 

As for the issues with the Multi Objective Loss, I will surely solve them and put a new PR, and we can experiment with it there.

Thankyou!

crsdvaibhav replied:
I have done the required changes: 
- Removed the Multi Objecive Loss 
- Resolved the conflicts

monatis reviewed (changes requested):
- `CosFace` LGTM, but let's remove `MultiObjectiveLoss`.
- Fix conflicts.

monatis commented:
I needed to close this. Could you please make another PR by rebasing from master and insuring that all the tests pass --because it didn't.
